File: 2021/src/main/python/day21.py
import numpy as np

#################################################
# Import
#################################################
with open("2021/src/main/resources/day21.txt", "r") as f:
    data = [int(row.strip()[-1]) for row in f.readlines()]

#################################################
# Prepare
#################################################
n = 10  # number of positions on board
n_end = 21  # number of points of end game

# Create dice transition matrix
row = np.eye(n, dtype=np.ulonglong)[0]
roll_mat = np.array([np.roll(row, i + 1) for i in range(n)], dtype=np.ulonglong)
roll_mat = roll_mat + roll_mat.dot(roll_mat) + roll_mat.dot(roll_mat).dot(roll_mat)
roll_mat = np.linalg.matrix_power(roll_mat, 3).T

# Create pawn move transition function
def move_pawn(pos: np.array) -> np.array:
    pos = pos.copy()
    for idx, row in enumerate(pos):
        temp = np.zeros(row.shape)
        temp[idx + 1:] = row[:-(idx + 1)]
        temp[-1] += row[-(idx + 1):].sum()
        pos[idx] = temp
    #
    return pos

# Prepare starting positions
pos = {0: np.zeros((n, n_end + 1), dtype=np.ulonglong), 1: np.zeros((n, n_end + 1), dtype=np.ulonglong)}
pos[0][data[0] - 1, 0] = 1
pos[1][data[1] - 1, 0] = 1

#################################################
# Process
#################################################
turn = 0
winning = np.zeros((1,2), dtype= np.ulonglong).flatten()
while all([p.sum() > 0 for p in pos.values()]):
    # Move pawn
    pos[turn] = move_pawn(roll_mat.dot(pos[turn]))
    # Check winnings
    winning[turn] += pos[turn][:, n_end].sum() * pos[1 - turn].sum()
    pos[turn][:, n_end] = 0
    # Go to next player
    turn = 1 - turn

#################################################
# Output
#################################################
print(f"{max(winning) = }")

File: 2021/scala/src/main/scala/Day21.scala
import scala.io.Source
import scala.annotation.tailrec
import scala.util.chaining._
import breeze.linalg._
import breeze.numerics._

@main def Day21 = {

  /** **********************************************
    * Import data
    * **********************************************
    */
  val input = Source.fromFile("2021/data/day21.txt").getLines.toList
  val data = input.map(_.last.asDigit)

  // Define dice roll transition matrix
  def createRollMat(n: Int): DenseMatrix[Long] =
    DenseMatrix
      .tabulate[Long](n, n) { (i, j) =>
        if ((10 + j - i + 3) % 10 < 3) 1 else 0
      }
      .pipe(r => r * r * r)

  // Apply moves of pawn base on dice roll
  def movePawn(p: DenseMatrix[Long]): (DenseMatrix[Long], Long) = {
    val n = p.cols
    val newPos = DenseMatrix.tabulate[Long](p.rows, p.cols) { (i, j) =>
      if (j > i && j < p.cols - 1) p(i, j - i - 1) else 0
    }
    val wins = (0 to p.rows - 1).map { i => sum(p(i, n - i - 2 to (n - 1))) }.sum
    (newPos, wins)
  }

  // Play game until end of all possibilities
  @tailrec
  def play(
    rollMat: DenseMatrix[Long]
  )(pos: Map[Int, DenseMatrix[Long]], turn: Int, winning: Seq[Long]): Seq[Long] =
    if sum(pos(0)) == 0 || sum(pos(1)) == 0 then winning
    else
      val (posTurn, newWins) = movePawn(rollMat * pos(turn))
      val nextWinning = Seq(winning(1), winning(0) + sum(pos(1 - turn)) * newWins)
      play(rollMat)(pos ++ Map(turn -> posTurn), 1 - turn, nextWinning)

  /** **********************************************
    * Prepare
    * **********************************************
    */
  // Playing game dimensions
  val (n, m) = (10, 21)

  // Starting points
  val pos = Map(
    0 -> DenseMatrix.tabulate[Long](n, m + 1) { (i, j) =>
      if (j == 0 && i == data(0) - 1) 1 else 0
    },
    1 -> DenseMatrix.tabulate[Long](n, m + 1) { (i, j) =>
      if (j == 0 && i == data(1) - 1) 1 else 0
    }
  )
  val rollMat = createRollMat(n)

  /** **********************************************
    * Process
    * **********************************************
    */
  val result = play(rollMat)(pos, 0, Seq(0, 0))

  /** **********************************************
    * Output
    * **********************************************
    */
  println(s"Max winning universe: ${result.max}")
}
